### PR TITLE
Bugfix. Inspect the logic of the data class rule

### LINKED ISSUE #750: Inspect the logic of the data class rule
## Describe the bug
```kotlin
class Credentials(auth: String) {
    val first = ...
    val second = ...

    init {
        ...
        foo(auth)
    }

...
}
```

The question is should it be considered as data class?

## Environment information
* diktat version: 0.4.0

### PR DESCRIPTION
**Pull request by aktsay6**
### What's done:
  * Logic extended

This pull request closes #750

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt [hunk 1]
```diff
@@ -56,15 +56,26 @@ class DataClassesRule(configRules: List<RulesConfig>) : DiktatRule("data-classes
     @Suppress("UnsafeCallOnNullableType", "FUNCTION_BOOLEAN_PREFIX", "ComplexMethod")
     private fun ASTNode.canBeDataClass(): Boolean {
         val isNotPropertyInClassBody = findChildByType(CLASS_BODY)?.let { (it.psi as KtClassBody).properties.isEmpty() } ?: true
+        val constructorParametersNames = mutableListOf<String>()
         val hasPropertyInConstructor = findChildByType(PRIMARY_CONSTRUCTOR)
             ?.let { constructor ->
                 (constructor.psi as KtPrimaryConstructor)
                     .valueParameters
-                    .run { isNotEmpty() && all { it.hasValOrVar() } }
+                    .run { isNotEmpty() && all { if (it.hasValOrVar()) true else { constructorParametersNames.add(it.name!!); false } } }
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt:64]: I'd move this `add` into separate `onEach` before `run`, now it's not very readable

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt [hunk 2]
```diff
@@ -56,15 +56,26 @@ class DataClassesRule(configRules: List<RulesConfig>) : DiktatRule("data-classes
...
             } ?: false
         if (isNotPropertyInClassBody && !hasPropertyInConstructor) {
             return false
         }
+        // if parameter of the primary constructor is used in init block then it is hard to refactor this class to data class
+        if (constructorParametersNames.isNotEmpty()) {
+            val initBlocks = findChildByType(CLASS_BODY)?.getAllChildrenWithType(CLASS_INITIALIZER)
+            initBlocks?.forEach { init ->
+                val refExpressions = init.findAllNodesWithSpecificType(REFERENCE_EXPRESSION)
+                if (refExpressions.any { it.text !in constructorParametersNames }) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt:74]: Is it right? Shouldn't it be 'any in'? Also please add tests with more parameters and more complex init blocks

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt [hunk 3]
```diff
@@ -56,15 +56,30 @@ class DataClassesRule(configRules: List<RulesConfig>) : DiktatRule("data-classes
...
             } ?: false
         if (isNotPropertyInClassBody && !hasPropertyInConstructor) {
             return false
         }
+        // if parameter of the primary constructor is used in init block then it is hard to refactor this class to data class
+        if (constructorParametersNames.isNotEmpty()) {
+            val initBlocks = findChildByType(CLASS_BODY)?.getAllChildrenWithType(CLASS_INITIALIZER)
+            initBlocks?.forEach { init ->
+                val refExpressions = init.findAllNodesWithSpecificType(REFERENCE_EXPRESSION)
+                if (refExpressions.any { it.text in constructorParametersNames }) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt:78]: Please add a test that would have passed when there was a mistake with `!in` and shouldn't actually pass.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt (hunk 1 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt]: Fixed

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt (hunk 2 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt]: Should be. Added more tests

**Review by petertrr — APPROVED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt (hunk 3 above)
Comment (aktsay6) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter6/classes/DataClassesRule.kt:78]: Added

**Pull request merged**